src/device.rs

```diff
@@ -2563,6 +2579,26 @@ fn set_tunnel_interface(socket_pool: &Arc<SocketPool>, config: &DeviceConfig) {
     }
 }
 
+fn node_from_exit_node(exit_node: &ExitNode) -> Node {
+    Node {
+        identifier: exit_node.identifier.clone(),
+        public_key: exit_node.public_key,
+        nickname: None,
+        is_exit: true,
+        is_vpn: exit_node.endpoint.is_some(),
+        ip_addresses: vec![IpAddr::V4(VPN_EXTERNAL_IPV4), IpAddr::V4(VPN_INTERNAL_IPV4)],
+        allowed_ips: exit_node.allowed_ips.clone().unwrap_or_default(),
+        endpoint: exit_node.endpoint,
+        hostname: None,
+        allow_incoming_connections: false,
+        allow_peer_send_files: false,
+        path: PathType::Direct,
+        allow_multicast: false,
+        peer_allows_multicast: false,
+        ..Default::default()
+    }
+}
+
```

Comment:
I wonder a bit about this code, a lot of fields have hardcoded values such as `allow_multicast: false` as an example. Is this the right conversion?

Nitpick: wouldn't it be more rustacean way to have a `From<T>` implemented and then use that instead?

Reply:
This is a helper function local to this module. This specific conversion was used, and I needed to make the same conversion somewhere else (within the module), so my idea was to extract this conversion into a helper function.
It isn't meant to be fully general though maybe it can be made so.

Regarding `From<>`
* both types are from a different crate (telio-model) - you cannot implement a trait on the foreign type, though it should be possible in `telio-model` crate
* as mentioned it wasn't meant to be general

So IDK how "correct" this conversion is so that it can be promoted to general and `From<>` be implemented. @Jauler any opinion on that?

Reply:
IMHO when implementing `From<X>` care must be taken to either:
* Make sure that transition is possible strictly into one direction (e.g. ExitNode -> Node, but not `Node -x-> ExitNode`. But you can always add more `From<T>` implementations, hence this is really hard to ensure.
or:
* Make sure that transition ExitNode -> Node -> ExitNode does not loose any information (note that the chain may be longer than two types).

Previously we had `From<T>` implemented for many pairs of types, especially various `Node` types, but that resulted in multitude bugs, where some unexpected transformation chain would loose some information. 

Therefore, if transformation is purpose-specific, like @matszczygiel mentioned -> we should explicitly signal that. Hence in my opinion purpose-specific helper function is suitable here and makes it less likely for unexpected transformation chains to form accidentally.

Reply:
Small correction: "you cannot implement a trait on the foreign type" is not true, but you cannot implement a *foreign* trait on a foreign type. It doesn't really change the discussion, just wanted to add the correction

Reply:
Good then. Leaving this as is, since this is very specific conversion